File: scripts/pack-fabric/routing/lib/customer-endpoint-access.js
"use strict";
const LIMIT_METERS = 200;
function customerEndpointEdges(pack, edgeIndex, seeds, reverse = false) {
  const edges = new Set();
  if (!(pack.graphBinaryVersion >= 4) || edgeIndex < 0 ||
      ![pack.edgeAccess[edgeIndex * 2], pack.edgeAccess[edgeIndex * 2 + 1]].includes(4)) return edges;
  const adjacency = new Map();
  for (let node = 0; node < pack.nodeCount; node++) {
    for (let arc = pack.nodeOffsets[node]; arc < pack.nodeOffsets[node + 1]; arc++) {
      const ei = pack.edgeUndirectedIndex[arc], to = pack.edgeTargets[arc];
      const code = pack.edgeAccess[ei * 2 + (pack.edgeFrom[ei] === node ? 0 : 1)];
      if (code !== 4) continue;
      const source = reverse ? to : node, target = reverse ? node : to;
      if (!adjacency.has(source)) adjacency.set(source, []);
      adjacency.get(source).push({to:target, edge:ei, meters:Number(pack.edgeMeters[ei])});
    }
  }
  const distances = new Map(), queue = [];
  for (const {node, meters} of seeds) {
    if (node < 0 || node >= pack.nodeCount || meters < 0 || meters > LIMIT_METERS || !Number.isFinite(meters)) continue;
    if (meters < (distances.get(node) ?? Infinity)) {
      distances.set(node, meters); queue.push({node, meters});
    }
  }
  if (queue.length) edges.add(edgeIndex);
  while (queue.length) {
    queue.sort((a,b)=>b.meters-a.meters);
    const current = queue.pop();
    if (current.meters !== distances.get(current.node)) continue;
    for (const arc of adjacency.get(current.node) || []) {
      const meters = current.meters + arc.meters;
      if (meters > LIMIT_METERS) continue;
      edges.add(arc.edge);
      if (meters < (distances.get(arc.to) ?? Infinity)) {
        distances.set(arc.to, meters); queue.push({node:arc.to,meters});
      }
    }
  }
  return edges;
}
// A search mask is only an admissible candidate set. Check the actual path:
// customer roads form an endpoint prefix/suffix, never a through shortcut,
// and actual travelled customer metres (not radial distance) obey the limit.
function validCustomerRuns(rows, customerIds, startCustomer, endCustomer) {
  for (let i=0;i<rows.length;) {
    if (!customerIds.has(String(rows[i].edgeId))) { i++; continue; }
    const first=i; let meters=0;
    while (i<rows.length && customerIds.has(String(rows[i].edgeId))) meters+=Number(rows[i++].meters)||0;
    if (meters>LIMIT_METERS+0.01 || !((first===0 && startCustomer) || (i===rows.length && endCustomer))) return false;
  }
  return true;
}
module.exports={customerEndpointEdges,validCustomerRuns,LIMIT_METERS};

File: Dirt/Routing/OnDevice/CustomerEndpointAccess.swift
import Foundation

nonisolated enum CustomerEndpointAccess {
    static let limitMeters = 200.0
    static func validRuns(_ rows: [(id: String, meters: Double)], customerIDs: Set<String>, start: Bool, end: Bool) -> Bool {
        var i = 0
        while i < rows.count {
            guard customerIDs.contains(rows[i].id) else { i += 1; continue }
            let first = i
            var meters = 0.0
            while i < rows.count, customerIDs.contains(rows[i].id) {
                meters += rows[i].meters
                i += 1
            }
            if meters > limitMeters + 0.01 || !((first == 0 && start) || (i == rows.count && end)) { return false }
        }
        return true
    }
}

extension GraphV2Pack {
    nonisolated func customerEndpointEdges(edgeIndex: Int, seeds: [(node: Int, meters: Double)], reverse: Bool = false) -> Set<Int> {
        guard version >= 4, legalTopology, edgeIndex >= 0,
              edgeIndex * 2 + 1 < edgeAccess.count,
              edgeAccess[edgeIndex * 2] == 4 || edgeAccess[edgeIndex * 2 + 1] == 4 else { return [] }
        var adjacency: [Int: [(to: Int, edge: Int, meters: Double)]] = [:]
        for node in 0..<nodeCount {
            for arc in Int(nodeOffsets[node])..<Int(nodeOffsets[node + 1]) {
                let ei = Int(edgeUndirectedIndex[arc])
                let target = Int(edgeTargets[arc])
                guard v4AccessCode(ei: ei, from: node, to: target) == 4 else { continue }
                let source = reverse ? target : node
                adjacency[source, default: []].append((reverse ? node : target, ei, Double(edgeMeters[ei])))
            }
        }
        var distances: [Int: Double] = [:]
        var queue: [(node: Int, meters: Double)] = []
        for seed in seeds where seed.node >= 0 && seed.node < nodeCount && seed.meters.isFinite && seed.meters >= 0 && seed.meters <= CustomerEndpointAccess.limitMeters {
            if seed.meters < (distances[seed.node] ?? .infinity) {
                distances[seed.node] = seed.meters
                queue.append(seed)
            }
        }
        var edges: Set<Int> = queue.isEmpty ? [] : [edgeIndex]
        while !queue.isEmpty {
            queue.sort { $0.meters > $1.meters }
            let current = queue.removeLast()
            guard current.meters == distances[current.node] else { continue }
            for arc in adjacency[current.node] ?? [] {
                let meters = current.meters + arc.meters
                guard meters <= CustomerEndpointAccess.limitMeters else { continue }
                edges.insert(arc.edge)
                if meters < (distances[arc.to] ?? .infinity) {
                    distances[arc.to] = meters
                    queue.append((arc.to, meters))
                }
            }
        }
        return edges
    }
}
